Update readme (#5)
collabora/ros-bridge: README.md

The ROS bridge package is a bridge that enables two-way communication between ROS and CARLA. The information from the CARLA server is translated to ROS topics and thus can be visualized in various visualization tools e.g. [FoxGlove Studio](https://github.com/foxglove/studio). This bridge can be used to setup and monitor a CARLA environment for simulated dataset generation for perception systems, image synthesis, and automatic ground truth data generation or testing the trained models.

> <img src="docs/images/layout.png" width="800"/>

Reviewer: Do you mind to center the image? -- example https://gist.github.com/DavidWells/7d2e0e1bc78f4ac59a123ddf8b74932d

Author: yeah sure. good idea!

> The ROS bridge package is a bridge that enables two-way communication between ROS and CARLA. The information from the CARLA server is translated to ROS topics and thus can be visualized in various visualization tools for e.g. [FoxGlove Studio](https://github.com/foxglove/studio). This bridge can be used to setup and monitor a CARLA environment for simulated dataset generation for perception systems, image synthesis, and automatic ground truth data generation or testing the trained models.

Reviewer: ```suggestion
The ROS bridge package is a bridge that enables two-way communication between ROS and CARLA. The information from the CARLA server is translated to ROS topics and thus can be visualized in various visualization tools e.g. [FoxGlove Studio](https://github.com/foxglove/studio). This bridge can be used to setup and monitor a CARLA environment for simulated dataset generation for perception systems, image synthesis, and automatic ground truth data generation or testing the trained models.

```

<img src="docs/images/demo1.png" width="800"/>

# What is ROS?

> The Robot Operating System (ROS) is a set of software libraries and tools that help you build robot applications. Most Robots are made up of actuators, sensors and control systems. ROS provides a standard way to visualize and connect all your sensors (cameras, distance sensors, analog to digital converters, IMU) together with the control software that makes the decisions.

Reviewer: ```suggestion
The Robot Operating System (ROS) is a set of software libraries and tools that help you build robot applications. Most Robots are made up of actuators, sensors, and control systems. ROS provides a standard way to visualize and connect all your sensors (cameras, distance sensors, analog to digital converters, IMU) together with the control software that makes the decisions.
```

> ROS helps developers quickly build these components and then easily connect them using ROS tools called topics and messages. These messages can be published on ROS topics and shared with visualization tools. Also, ROS can be used with simulated robots which makes training and testing easier, faster and more effective.
> 

Reviewer: ```suggestion
ROS helps developers quickly build these components and easily connect them using ROS tools called topics and messages. These messages can be published on ROS topics and shared with visualization tools. Also, ROS can be used with simulated robots, which makes training and testing easier, faster, and more effective.
```

# ROS with CARLA

> CARLA is an open-source simulator for autonomous driving research. The simulator provides API to spawn vehicles into the CARLA world which can be controlled by custom models or via autopilot. Moreover, it comes with various maps of different environments like country side roads or urban areas. It also provides methods to change weather conditions like time of day, precipitation etc.

Reviewer: ```suggestion
CARLA is an open-source simulator for autonomous driving research. The simulator provides API to spawn vehicles into the CARLA world, which can be controlled by custom models or via autopilot. Moreover, it comes with various maps of different environments like countryside roads or urban areas. It also provides methods to change weather conditions like time of day, precipitation, etc.

```

## ROS bridge Sensors & our additions

> CARLA simulator provides methods to attach multiple sensors to CARLA vehicles(actors) like RGB cameras, semantic segmentation camera, depth camera, IMU, LiDAR, semantic LiDAR. It also allows to simulate custom traffic scenarios.

Reviewer: ```suggestion
CARLA simulator provides methods to attach multiple sensors to CARLA vehicles(actors) like RGB cameras, semantic segmentation camera, depth camera, IMU, LiDAR, semantic LiDAR. It also allows simulating custom traffic scenarios.

```

This repository is a fork of carla-simulator/ros-bridge which already had various sensor data converted to ROS messages, some of which we had to modify as well to fit our usecases:

- RGB Camera - [sensor_msgs/CompressedImage](https://docs.ros.org/en/api/sensor_msgs/html/msg/CompressedImage.html)
- Depth Camera - [sensor_msgs/CompressedImage](https://docs.ros.org/en/api/sensor_msgs/html/msg/CompressedImage.html)
- Semantic Segmentation Camera - [sensor_msgs/CompressedImage](https://docs.ros.org/en/api/sensor_msgs/html/msg/CompressedImage.html)
- Vehicle Position - [geometry_msgs/PoseStamped](http://docs.ros.org/en/noetic/api/geometry_msgs/html/msg/PoseStamped.html), [sensor_msgs/NavSatFix](https://docs.ros.org/en/api/sensor_msgs/html/msg/NavSatFix.html), [tf2_msgs/TFMessage](http://docs.ros.org/en/noetic/api/tf2_msgs/html/msg/TFMessage.html)
- Velocity and Acceleration - [carla_msgs/CarlaEgoVehicleStatus](https://github.com/carla-simulator/ros-carla-msgs/blob/master/msg/CarlaEgoVehicleStatus.msg)

> Apart from that we added some new features and updates:

Reviewer: ```suggestion
Apart from that, we added some new features and updates:

```

- Converted all the camera message types from ```sensor_msgs/Image``` to ```sensor_msgs/CompressedImage```.
- Added support for ```ros map_server``` for ROS1 and ```nav2_map_server map_server``` for ROS2 to convert map images to ```nav_msgs/OccupancyGrid``` so that the map is able to be displayed on 3D panel in visualization tool like FoxGlove Studio or WebViz.

> - Added PCL maps on top of ```ROS1/ROS2 map_server``` to display more detailed maps on the 3D panel of your favourite visualization tool.

Reviewer: ```suggestion
- Added PCL maps on top of ```ROS1/ROS2 map_server``` to display more detailed maps on the 3D panel of your favorite visualization tool.

```

- Added support to publish ego vehicle traversed path as ```nav_msgs/Path``` which itself is a list of waypoints ```geometry_msgs/PoseStamped```.
- Added a separate ROS topic that publishes RGB camera images with 3D Bounding boxes and LiDAR points overlaid.

# How to use

> Autonomous vehicles require large-scale development and testing in a wide range of scenarios before they can be deployed. Moreover, huge amounts of data is required to train models for planning and perception which can be costly and time consuming. To solve the problem, simulated datasets from the CARLA simulater can be a great solution but still could be time consuming and might require diligence to validate the data gathering process.

Reviewer: ```suggestion
Autonomous vehicles require large-scale development and testing in a wide range of scenarios before they can be deployed. Moreover, huge amounts of data are required to train models for planning and perception, which can be costly and time consuming. To solve the problem, simulated datasets from the CARLA simulator can be a great solution but still could be time-consuming and might require diligence to validate the data gathering process.

```

## CarlaFox

> We implemented [Carlafox](https://github.com/collabora/carlafox) - a web-based CARLA visualizer that can combine a number of ROS topics message streams including support for text and custom shapes in one single visualization to solve the problem of monitoring the data gathering process or testing your perception/planning models.

Reviewer: ```suggestion
We implemented [Carlafox](https://github.com/collabora/carlafox) - a web-based CARLA visualizer that can combine a number of ROS topics message streams, including support for text and custom shapes in one single visualization to solve the problem of monitoring the data gathering process or testing your perception/planning models.

```

> Carlafox eradicates the hassle of using the ros-bridge with CARLA and adds FoxGlove Visualization on top of that. Instead of following intricate steps like running CARLA server, starting the ros-bridge and finally, starting FoxGlove or any other visualization tool. Carlafox pulls everything together and make it a lot easier to get started.

Reviewer: ```suggestion
Carlafox eradicates the hassle of using the ros-bridge with CARLA and adds FoxGlove Visualization on top of that. Instead of following intricate steps like running CARLA server, starting the ros-bridge, and finally, starting FoxGlove or any other visualization tool. Carlafox pulls everything together and makes it a lot easier to get started.

```

The ROS bridge package is a bridge that enables two-way communication between ROS and CARLA. The information from the CARLA server is translated to ROS topics and thus can be visualized in various visualization tools e.g. [FoxGlove Studio](https://github.com/foxglove/studio). This bridge can be used to setup and monitor a CARLA environment for simulated dataset generation for perception systems, image synthesis, and automatic ground truth data generation or testing the trained models.

> <img src="docs/images/demo1.png" width="800"/>

Reviewer: Can you update the image and use the new map.

> 

Reviewer: ```suggestion
```

Remove the extra line.

# What is ROS?
The Robot Operating System (ROS) is a set of software libraries and tools that help you build robot applications. Most Robots are made up of actuators, sensors, and control systems. ROS provides a standard way to visualize and connect all your sensors (cameras, distance sensors, analog to digital converters, IMU) together with the control software that makes the decisions.

ROS helps developers quickly build these components and easily connect them using ROS tools called topics and messages. These messages can be published on ROS topics and shared with visualization tools. Also, ROS can be used with simulated robots, which makes training and testing easier, faster, and more effective.

# ROS with CARLA
CARLA is an open-source simulator for autonomous driving research. The simulator provides API to spawn vehicles into the CARLA world, which can be controlled by custom models or via autopilot. Moreover, it comes with various maps of different environments like countryside roads or urban areas. It also provides methods to change weather conditions like time of day, precipitation, etc.

> 

Reviewer: ```suggestion
```

Remove the extra line.

- Added support for ```ros map_server``` for ROS1 and ```nav2_map_server map_server``` for ROS2 to convert map images to ```nav_msgs/OccupancyGrid``` so that the map is able to be displayed on 3D panel in visualization tool like FoxGlove Studio or WebViz.
- Added PCL maps on top of ```ROS1/ROS2 map_server``` to display more detailed maps on the 3D panel of your favorite visualization tool.
- Added support to publish ego vehicle traversed path as ```nav_msgs/Path``` which itself is a list of waypoints ```geometry_msgs/PoseStamped```.
- Added a separate ROS topic that publishes RGB camera images with 3D Bounding boxes and LiDAR points overlaid.

# How to use
Autonomous vehicles require large-scale development and testing in a wide range of scenarios before they can be deployed. Moreover, huge amounts of data are required to train models for planning and perception, which can be costly and time consuming. To solve the problem, simulated datasets from the CARLA simulator can be a great solution but still could be time-consuming and might require diligence to validate the data gathering process.

> 

Reviewer: ```suggestion
```

Remove the extra line.

## CarlaFox
We implemented [Carlafox](https://github.com/collabora/carlafox) - a web-based CARLA visualizer that can combine a number of ROS topics message streams, including support for text and custom shapes in one single visualization to solve the problem of monitoring the data gathering process or testing your perception/planning models.

> 

Reviewer: ```suggestion
```

Remove the extra line.

Carlafox eradicates the hassle of using the ros-bridge with CARLA and adds FoxGlove Visualization on top of that. Instead of following intricate steps like running CARLA server, starting the ros-bridge, and finally, starting FoxGlove or any other visualization tool. Carlafox pulls everything together and makes it a lot easier to get started.

> 

Reviewer: ```suggestion
```

Remove the extra line.

## ROS Bridge & CARLA
Besides, you can also use the ros-bridge without Carlafox. 
- Clone the ros-bridge repo with carla_msgs submodule.
```bash
 git clone --recurse-submodules https://github.com/makaveli10/ros-bridge.git
 cd ros-bridge/
```

- Build docker image with ROS1/ROS2
```bash

>  docker build -t carlafox . --build-arg ROS_DISTRO=noetic
>                                         ROS_DISTRO=foxy

Reviewer: ```suggestion
 docker build -t carlafox . --build-arg ROS_DISTRO=noetic ROS_DISTRO=foxy
```

So that GitHub can render this in a single line and it's easier to copy/paste the command.
